In [1]:
import pandas as pd

In [2]:
# this csv file is 379 MB and too large to place on Github
# in order to run this notebook, you must first download the file and place on your local machine in the data/ folder
# the csv file is available here: https://www.microns-explorer.org/phase1 (scroll down to synapse table section)

synapses = pd.read_csv("pni_synapses_v185.csv")

In [3]:
synapses

id         pre_root_id        post_root_id  cleft_vx  \
0        2715757  648518346349538285  648518346349536759      1103   
1          12498  648518346342404863  648518346342414574        62   
2          12500  648518346342921567  648518346345323296        62   
3          12517  648518346342796822  648518346349537255       955   
4          12503  648518346342800412  648518346348190542       718   
...          ...                 ...                 ...       ...   
3239270  2865917  648518346348045782  648518346349537840       106   
3239271  3386018  648518346347834064  648518346349537840        50   
3239272  2940412  648518346347876810  648518346349537840       139   
3239273  2682848  648518346348343883  648518346349537840       987   
3239274  2120395  648518346348525239  648518346349537840       387   

         ctr_pt_x_nm  ctr_pt_y_nm  ctr_pt_z_nm  pre_pos_x_vx  pre_pos_y_vx  \
0             280784       261896        40600         70050         65534   
1             365832       292808        25800         91472         73196   
2             229144       155952        48080         57250         39028   
3             398452       208560        50760         99538         52184   
4             461868       245212        53560        115462         61350   
...              ...          ...          ...           ...           ...   
3239270       268776       173356        14360         67190         43302   
3239271       267344       156520         8520         66762         39152   
3239272       292772       156680        12600         73136         39170   
3239273       462768       237776         8520        115680         59370   
3239274       461000       237944         8920        115270         59526   

         pre_pos_z_vx  ctr_pos_x_vx  ctr_pos_y_vx  ctr_pos_z_vx  \
0                1015         70196         65474          1015   
1                 645         91458         73202           645   
2                1202         57286         38988          1202   
3                1276         99613         52140          1269   
4                1335        115467         61303          1339   
...               ...           ...           ...           ...   
3239270           361         67194         43339           359   
3239271           213         66836         39130           213   
3239272           315         73193         39170           315   
3239273           208        115692         59444           213   
3239274           222        115250         59486           223   

         post_pos_x_vx  post_pos_y_vx  post_pos_z_vx  
0                70104          65602           1015  
1                91396          73216            645  
2                57312          38966           1203  
3                99550          52098           1276  
4               115468          61296           1335  
...                ...            ...            ...  
3239270          67216          43364            360  
3239271          66874          39134            213  
3239272          73218          39160            315  
3239273         115628          59430            214  
3239274         115200          59464            222  

[3239275 rows x 16 columns]

In [4]:
# query: what cell has the most pre-synaptic sites?
# First, do a value count on pre_root_id
synapses["pre_root_id"].value_counts().max()

1475

In [5]:
# Next, find the cellid for this cell
mostpre = synapses["pre_root_id"].value_counts()

In [6]:
mostpre.to_frame("pre_count").head(10)

pre_count
648518346349539215       1475
648518346341376986       1193
648518346349526102       1169
648518346349528994       1082
648518346349516216        769
648518346349539846        761
648518346349514408        715
648518346349529031        713
648518346349531742        685
648518346349515607        676

In [7]:
# query: what cell has the most post-synaptic sites?
# First, do a value count on post_root_id
synapses["post_root_id"].value_counts().max()

5369

In [8]:
mostpost = synapses["post_root_id"].value_counts()

In [9]:
mostpost.to_frame("post_count").head(10)

post_count
648518346349538440        5369
648518346349540055        5191
648518346349539856        5170
648518346349539819        5109
648518346349539896        4935
648518346349539215        4891
648518346349538426        4886
648518346349528994        4831
648518346349517783        4818
648518346349539785        4785

In [10]:
# define a variable for this cellid

cellidlookup0 = 648518346349538440

This is the excitatory pyramidal neuron with the most post-synaptic sites in the volume;  
Identified using the cellid lookup table, for example use this notebook to look up cell subtypes:  
https://github.com/shandran/layer23-volume/blob/main/lookup_cellid_in_layer23_volume.ipynb

In [11]:
query0pre = synapses[synapses["pre_root_id"] == cellidlookup0]
query0post = synapses[synapses["post_root_id"] == cellidlookup0]

In [12]:
len(query0pre)

106

This excitatory pyramidal neuron has 106 synapses located in its axon (hence these are pre-synaptic sites);  
In other words, this neuron is making 106 synapses onto other processes in the volume;  
(these are all making synaptic contacts with processes that are contained within the volume)

In [13]:
len(query0post)

5369

This excitatory pyramidal neuron has 5369 synapses in its dendrites and soma;  
In other words, this neuron is receiving 5369 synapses from other processes;  
(these could be from neurons located inside or outside the volume)

In [14]:
query0pre

id         pre_root_id        post_root_id  cleft_vx  \
43385    3362441  648518346349538440  648518346349533227        93   
77607     585587  648518346349538440  648518346349533719       112   
87444     741207  648518346349538440  648518346349532796       128   
98677       7163  648518346349538440  648518346349517783        62   
108599   2356651  648518346349538440  648518346349537153       523   
...          ...                 ...                 ...       ...   
3125154  2274529  648518346349538440  648518346349528994       468   
3125157  3099379  648518346349538440  648518346349537404        60   
3125158  2384556  648518346349538440  648518346349537583       547   
3125161   305572  648518346349538440  648518346349539863        57   
3125398  2703875  648518346349538440  648518346341377469      1062   

         ctr_pt_x_nm  ctr_pt_y_nm  ctr_pt_z_nm  pre_pos_x_vx  pre_pos_y_vx  \
43385         412616       220784        80360        103128         55178   
77607         445976       218688        22440        111468         54626   
87444         412896       221336        76920        103258         55358   
98677         352248       262368        60440         88004         65592   
108599        408096       249616        53000        101964         62378   
...              ...          ...          ...           ...           ...   
3125154       394216       262720        57840         98564         65648   
3125157       374300       250268        74880         93578         62548   
3125158       423872       238624        55160        105982         59698   
3125161       328248       227888        59160         82020         56974   
3125398       231992       271792        54680         57920         67960   

         pre_pos_z_vx  ctr_pos_x_vx  ctr_pos_y_vx  ctr_pos_z_vx  \
43385            2010        103154         55196          2009   
77607             561        111494         54672           561   
87444            1925        103224         55334          1923   
98677            1510         88062         65592          1511   
108599           1321        102024         62404          1325   
...               ...           ...           ...           ...   
3125154          1446         98554         65680          1446   
3125157          1872         93575         62567          1872   
3125158          1379        105968         59656          1379   
3125161          1480         82062         56972          1479   
3125398          1369         57998         67948          1367   

         post_pos_x_vx  post_pos_y_vx  post_pos_z_vx  
43385           103174          55202           2008  
77607           111512          54686            562  
87444           103200          55308           1925  
98677            88106          65582           1510  
108599          102054          62402           1321  
...                ...            ...            ...  
3125154          98474          65664           1446  
3125157          93656          62570           1871  
3125158         105922          59650           1379  
3125161          82082          56966           1480  
3125398          57952          67888           1369  

[106 rows x 16 columns]

In [15]:
query0post

id         pre_root_id        post_root_id  cleft_vx  \
14268     812010  648518346349535192  648518346349538440       135   
843541   1064600  648518346349536679  648518346349538440       163   
1673868  2128208  648518346349535192  648518346349538440       391   
2108218   961438  648518346349538218  648518346349538440       151   
2108223   227805  648518346349537711  648518346349538440        80   
...          ...                 ...                 ...       ...   
3158194   308829  648518346342795183  648518346349538440        87   
3158195   308690  648518346346310829  648518346349538440        87   
3158196   193890  648518346347712242  648518346349538440      2053   
3158197   308397  648518346341411847  648518346349538440        87   
3158198   265571  648518346342798532  648518346349538440        83   

         ctr_pt_x_nm  ctr_pt_y_nm  ctr_pt_z_nm  pre_pos_x_vx  pre_pos_y_vx  \
14268         407712       226336        49080        101922         56558   
843541        295960       218744        35000         73942         54666   
1673868       366128       242728        16480         91532         60636   
2108218       391184       201976        23560         97782         50514   
2108223       396640       252088        42760         99170         63068   
...              ...          ...          ...           ...           ...   
3158194       365096       196568        66080         91258         49086   
3158195       290976       224360        58440         72730         56068   
3158196       307392       157948        23960         76762         39500   
3158197       367744       248888        73480         91888         62198   
3158198       374824       272128        81720         93722         67982   

         pre_pos_z_vx  ctr_pos_x_vx  ctr_pos_y_vx  ctr_pos_z_vx  \
14268            1226        101928         56584          1227   
843541            876         73990         54686           875   
1673868           415         91532         60682           412   
2108218           590         97796         50494           589   
2108223          1070         99160         63022          1069   
...               ...           ...           ...           ...   
3158194          1652         91274         49142          1652   
3158195          1459         72744         56090          1461   
3158196           598         76848         39487           599   
3158197          1834         91936         62222          1837   
3158198          2043         93706         68032          2043   

         post_pos_x_vx  post_pos_y_vx  post_pos_z_vx  
14268           101922          56612           1226  
843541           74022          54688            877  
1673868          91570          60706            416  
2108218          97820          50476            590  
2108223          99128          63004           1069  
...                ...            ...            ...  
3158194          91294          49166           1652  
3158195          72764          56118           1459  
3158196          76774          39424            597  
3158197          91972          62204           1834  
3158198          93684          68066           2043  

[5369 rows x 16 columns]

In [16]:
#export a list of synapse ids in the pre-synaptic query; e.g., useful for copying and pasting into neuroglancer
# this is a long list of 106 synapse ids; remove the # from the line below to run

# query0pre["id"].to_list()

In [17]:
#export a list of synapse ids in the post-synaptic query; e.g., useful for copying and pasting into neuroglancer
# this is a really long list 5469 synapse ids; remove the # from the line below to run

# query0post["id"].to_list()

In [18]:
# repeat for the neuron with the most pre-synaptic sites in the volume 

cellidlookup1 = 648518346349539215

Using the cellid lookup notebook, this is an inhibitory basket neuron

In [19]:
query1pre = synapses[synapses["pre_root_id"] == cellidlookup1]
query1post = synapses[synapses["post_root_id"] == cellidlookup1]

In [20]:
len(query1pre)

1475

In [21]:
len(query1post)

4891

In [22]:
query1pre

id         pre_root_id        post_root_id  cleft_vx  \
10588    1600812  648518346349539215  648518346349537649       243   
32022    1047036  648518346349539215  648518346349537687       887   
32023    2478455  648518346349539215  648518346349537687       635   
32039     517090  648518346349539215  648518346349537687       106   
34953    2437346  648518346349539215  648518346349538746       597   
...          ...                 ...                 ...       ...   
3215466  1027398  648518346349539215  648518346349538209       159   
3220206  2228186  648518346349539215  648518346349531742       441   
3220560  1473317  648518346349539215  648518346349538793       221   
3225227   999623  648518346349539215  648518346349538730       155   
3238895  1186253  648518346349539215  648518346349536924       178   

         ctr_pt_x_nm  ctr_pt_y_nm  ctr_pt_z_nm  pre_pos_x_vx  pre_pos_y_vx  \
10588         438296       277464         3720        109580         69282   
32022         404988       157180        68400        101224         39444   
32023         393760       170160        74440         98366         42528   
32039         359768       200320        12520         89950         50020   
34953         354808       299056        80800         88696         74818   
...              ...          ...          ...           ...           ...   
3215466       264168       211952        60640         65996         52956   
3220206       381536       199680        79280         95310         49898   
3220560       367944       183168        26000         91932         45842   
3225227       260800       209024        25560         65242         52226   
3238895       381816       196112        75400         95406         49052   

         pre_pos_z_vx  ctr_pos_x_vx  ctr_pos_y_vx  ctr_pos_z_vx  \
10588              95        109574         69366            93   
32022            1709        101247         39295          1710   
32023            1866         98440         42540          1861   
32039             312         89942         50080           313   
34953            2016         88702         74764          2020   
...               ...           ...           ...           ...   
3215466          1517         66042         52988          1516   
3220206          1983         95384         49920          1982   
3220560           650         91986         45792           650   
3225227           637         65200         52256           639   
3238895          1888         95454         49028          1885   

         post_pos_x_vx  post_pos_y_vx  post_pos_z_vx  
10588           109516          69342             95  
32022           101122          39426           1709  
32023            98466          42506           1866  
32039            89886          50078            312  
34953            88656          74726           2016  
...                ...            ...            ...  
3215466          66050          53034           1517  
3220206          95406          49874           1983  
3220560          91918          45754            651  
3225227          65160          52250            636  
3238895          95478          49032           1889  

[1475 rows x 16 columns]

In [23]:
query1post

id         pre_root_id        post_root_id  cleft_vx  \
33788    2261810  648518346349536947  648518346349539215       461   
33791    2072778  648518346349536947  648518346349539215       369   
42521    2504052  648518346349537840  648518346349539215       667   
42525    2438498  648518346349537840  648518346349539215       594   
45942     122460  648518346347455240  648518346349539215        71   
...          ...                 ...                 ...       ...   
3126111  2183307  648518346349537848  648518346349539215       419   
3126408   738359  648518346349539071  648518346349539215       127   
3126409  2064591  648518346349539071  648518346349539215       366   
3126713  1667100  648518346349538527  648518346349539215       256   
3126714  1006064  648518346349538527  648518346349539215       156   

         ctr_pt_x_nm  ctr_pt_y_nm  ctr_pt_z_nm  pre_pos_x_vx  pre_pos_y_vx  \
33788         426144       260376        78120        106564         65080   
33791         437280       276960        68880        109344         69214   
42521         423552       271072        54640        105924         67820   
42525         419344       268608         4400        104782         67076   
45942         357704       212088         9720         89408         53060   
...              ...          ...          ...           ...           ...   
3126111       407016       242136         4080        101822         60492   
3126408       422824       269840        50600        105724         67496   
3126409       423040       253088        65440        105710         63210   
3126713       407208       238336        32320        101828         59568   
3126714       404992       237312        31200        101252         59284   

         pre_pos_z_vx  ctr_pos_x_vx  ctr_pos_y_vx  ctr_pos_z_vx  \
33788            1949        106536         65094          1953   
33791            1722        109320         69240          1722   
42521            1362        105888         67768          1366   
42525             111        104836         67152           110   
45942             244         89426         53022           243   
...               ...           ...           ...           ...   
3126111           104        101754         60534           102   
3126408          1263        105706         67460          1265   
3126409          1637        105760         63272          1636   
3126713           808        101802         59584           808   
3126714           779        101248         59328           780   

         post_pos_x_vx  post_pos_y_vx  post_pos_z_vx  
33788           106468          65084           1949  
33791           109256          69246           1722  
42521           105886          67718           1362  
42525           104776          67166            111  
45942            89340          52998            244  
...                ...            ...            ...  
3126111         101780          60580            104  
3126408         105648          67442           1263  
3126409         105760          63302           1637  
3126713         101782          59616            808  
3126714         101228          59384            779  

[4891 rows x 16 columns]

In [24]:
#export a list of synapse ids in the pre-synaptic query; e.g., useful for copying and pasting into neuroglancer
# this is a really long list of 1475 ids; remove the # from the line below to run

# query1pre["id"].to_list()

In [25]:
#export a list of synapse ids in the post-synaptic query; e.g., useful for copying and pasting into neuroglancer
# this is a really long list 4891 synapse ids; remove the # from the line below to run

# query1post["id"].to_list()

Working copy of the query function

In [26]:
# enter a cell id of interest
# e.g., select from the synapse table, or use the copy function in neuroglancer in the cell segmentation panel

cellidlookup = 648518346349540055

In [27]:
query_pre = synapses[synapses["pre_root_id"] == cellidlookup]
query_post = synapses[synapses["post_root_id"] == cellidlookup]

In [28]:
len(query_pre)

9

In [29]:
len(query_post)

5191

In [30]:
query_pre

id         pre_root_id        post_root_id  cleft_vx  \
6968       27729  648518346349540055  648518346344132118        63   
115128    184347  648518346349540055  648518346349380848        76   
159192     20073  648518346349540055  648518346341366757        63   
261372    208743  648518346349540055  648518346342801297        78   
1531428    29825  648518346349540055  648518346349539591        64   
2511006  3160775  648518346349540055  648518346349378902        56   
2706511  3404434  648518346349540055  648518346344072312        51   
2747161  3555058  648518346349540055  648518346342798924        61   
2796321  3528814  648518346349540055  648518346349517783        59   

         ctr_pt_x_nm  ctr_pt_y_nm  ctr_pt_z_nm  pre_pos_x_vx  pre_pos_y_vx  \
6968          436512       289512         9920        109138         72390   
115128        384040       163216        28640         96026         40802   
159192        365504       218640        67960         91316         54640   
261372        335992       241080        54520         83938         60258   
1531428       303664       286056        72240         75850         71498   
2511006       393940       187296        63120         98440         46856   
2706511       422208       281032        19480        105510         70218   
2747161       308136       253304        45080         77020         63332   
2796321       322696       287400        31080         80688         71806   

         pre_pos_z_vx  ctr_pos_x_vx  ctr_pos_y_vx  ctr_pos_z_vx  \
6968              249        109128         72378           248   
115128            717         96010         40804           716   
159192           1701         91376         54660          1699   
261372           1362         83998         60270          1363   
1531428          1808         75916         71514          1806   
2511006          1578         98485         46824          1578   
2706511           486        105552         70258           487   
2747161          1128         77034         63326          1127   
2796321           777         80674         71850           777   

         post_pos_x_vx  post_pos_y_vx  post_pos_z_vx  
6968            109090          72354            248  
115128           95990          40860            716  
159192           91416          54640           1701  
261372           84036          60236           1362  
1531428          75940          71486           1807  
2511006          98500          46804           1579  
2706511         105572          70280            486  
2747161          77034          63360           1127  
2796321          80640          71896            777

In [31]:
query_post

id         pre_root_id        post_root_id  cleft_vx  \
595856    832348  648518346349536810  648518346349540055       137   
595858   1904449  648518346349536935  648518346349540055       607   
595865   1801857  648518346349536888  648518346349540055       286   
595888   3018417  648518346349536935  648518346349540055       299   
596025    923429  648518346349536935  648518346349540055       147   
...          ...                 ...                 ...       ...   
3162653  1805437  648518346342789779  648518346349540055       676   
3162654  2187461  648518346349538691  648518346349540055       420   
3162658  1383592  648518346346305951  648518346349540055       206   
3162662   774287  648518346349508268  648518346349540055       131   
3162663   737083  648518346349536935  648518346349540055       127   

         ctr_pt_x_nm  ctr_pt_y_nm  ctr_pt_z_nm  pre_pos_x_vx  pre_pos_y_vx  \
595856        310488       278080        18560         77568         69556   
595858        339744       254568        20360         84858         63690   
595865        422680       307040        68480        105608         76736   
595888        344244       250220        19360         86092         62526   
596025        343072       216592        13920         85746         54126   
...              ...          ...          ...           ...           ...   
3162653       302904       281604        68400         75690         70408   
3162654       344272       267656        32200         86060         66848   
3162658       306504       235760        31560         76598         58954   
3162662       334464       172640        18000         83644         43142   
3162663       348096       267496        21280         86954         66878   

         pre_pos_z_vx  ctr_pos_x_vx  ctr_pos_y_vx  ctr_pos_z_vx  \
595856            463         77622         69520           464   
595858            502         84936         63642           509   
595865           1709        105670         76760          1712   
595888            483         86061         62555           484   
596025            350         85768         54148           348   
...               ...           ...           ...           ...   
3162653          1715         75726         70401          1710   
3162654           803         86068         66914           805   
3162658           788         76626         58940           789   
3162662           449         83616         43160           450   
3162663           530         87024         66874           532   

         post_pos_x_vx  post_pos_y_vx  post_pos_z_vx  
595856           77630          69502            462  
595858           84818          63600            502  
595865          105702          76764           1709  
595888           85988          62552            483  
596025           85766          54176            349  
...                ...            ...            ...  
3162653          75616          70336           1714  
3162654          85978          66888            803  
3162658          76618          58900            788  
3162662          83580          43186            449  
3162663          87056          66904            530  

[5191 rows x 16 columns]

In [32]:
query_pre["id"].to_list()

[27729, 184347, 20073, 208743, 29825, 3160775, 3404434, 3555058, 3528814]

In [33]:
# commented out due to large size of list
# query_post["id"].to_list()

Visualize neurons of interest using meshparty and vtk using this example notebook:  
https://github.com/shandran/layer23-volume/blob/main/most_synapses/neurons_with_most_synapses_layer23.ipynb